bindings/nodejs/src/functions/oic-client-discovery.cc

```
@@ -0,0 +1,100 @@
+    bool keepDiscovering = !!resource;
+    OicCallbackData *callbackData = (OicCallbackData *)data;
+
+    // Call the JS callback
+    Local<Value> arguments[2] = {
+        Nan::New(*(callbackData->jsClient)),
+        Nan::Null()
+    };
+    if (resource) {
+        arguments[1] = SolOicClientResource::New(resource);
+    }
+    Local<Value> jsResult = callbackData->callback->Call(2, arguments);
+
+    // Determine whether we should keep discovering
+    if (!jsResult->IsBoolean()) {
+        Nan::ThrowTypeError(
+            "Resource discovery callback return value is not boolean");
+    } else {
+        keepDiscovering = keepDiscovering &&
+            Nan::To<bool>(jsResult).FromJust();
+    }
+
+    // Tear down if discovery is done
+    if (!keepDiscovering) {
+        delete callbackData;
+    }
+
+    return keepDiscovering;
+}
+
+NAN_METHOD(bind_sol_oic_client_find_resource) {
+    VALIDATE_ARGUMENT_COUNT(info, 4);
+    VALIDATE_ARGUMENT_TYPE(info, 0, IsObject);
+    VALIDATE_ARGUMENT_TYPE(info, 1, IsObject);
+    VALIDATE_ARGUMENT_TYPE(info, 2, IsString);
+    VALIDATE_ARGUMENT_TYPE(info, 3, IsFunction);
+
+    struct sol_network_link_addr theAddress;
+    if (!c_sol_network_link_addr(Nan::To<Object>(info[1]).ToLocalChecked(),
+        &theAddress)) {
+        return;
+    }
+
+    Local<Object> jsClient = Nan::To<Object>(info[0]).ToLocalChecked();
+    struct sol_oic_client *client = (struct sol_oic_client *)
+        SolOicClient::Resolve(jsClient);
+    if (!client) {
+        return;
+    }
+
+    OicCallbackData *callbackData =
+        OicCallbackData::New(jsClient, Local<Function>::Cast(info[3]));
+    if (!callbackData) {
+        return;
+    }
+
+    bool result = sol_oic_client_find_resource((struct sol_oic_client *)client,
+        &theAddress, (const char *)*String::Utf8Value(info[2]), resourceFound,
+        callbackData);
+
+    if (!result) {
+        delete callbackData;
+    }
+
+    info.GetReturnValue().Set(Nan::New(result));
+}
```

The signature of a JS binding returns void, but accepts a parameter, ```info```, which contains a return value which, in turn, can be set. This way, if you need to throw a JS exception, you can just throw it and return. That plays the role of the JS jump that happens when an exception is thrown. The effect of the above code is that ```result``` becomes the return value of the binding.